xrootd/client/protocol.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package client // import "go-hep.org/x/hep/xrootd/client"
+
+import (
+	"context"
+
+	"go-hep.org/x/hep/xrootd/encoder"
+	"go-hep.org/x/hep/xrootd/protocol"
+	protocolRequest "go-hep.org/x/hep/xrootd/protocol/protocol"
+)
+
+// ProtocolInfo is a response for the `Protocol` request. See details in the xrootd protocol specification.
+type ProtocolInfo struct {
+	BinaryProtocolVersion int32
+	ServerType            protocol.ServerType
+	IsManager             bool
+	IsServer              bool
+	IsMeta                bool
+	IsProxy               bool
+	IsSupervisor          bool
+	SecurityVersion       byte
+	ForceSecurity         bool
+	SecurityLevel         byte
+	SecurityOverrides     []protocolRequest.SecurityOverride
+}
+
+/*
+Protocol obtains the protocol version number, type of the server and security information, such as:
+security version, security options, security level, and list of alterations
+needed to the specified predefined security level.
+*/
+func (client *Client) Protocol(ctx context.Context) (ProtocolInfo, error) {
+	resp, err := client.call(ctx, protocolRequest.RequestID, protocolRequest.NewRequest(client.protocolVersion, true))
+	if err != nil {
+		return ProtocolInfo{}, err
+	}
+
+	var generalResp protocolRequest.GeneralResponse
+
+	if err = encoder.Unmarshal(resp, &generalResp); err != nil {
+		return ProtocolInfo{}, err
+	}
+
+	var securityInfo *protocolRequest.SecurityInfo
+	if len(resp) > protocolRequest.GeneralResponseLength {
+		securityInfo = &protocolRequest.SecurityInfo{}
+		err = encoder.Unmarshal(resp[protocolRequest.GeneralResponseLength:], securityInfo)
+	}
+
+	var info = ProtocolInfo{}
```

Comment:
it seems we could create a `ProtocolInfo` value with much of its fields already initialized to the correct values, like so:

```go
info := ProtocolInfo{
    BinaryProtocolVersion: generalResp.BinaryProtocolVersion,
    ServerType: extractServerType(generalResp.Flags),
    ...
}
```